### PR TITLE
Use new attestation protection

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
This pr is closing the loop and makes use of the new attestation protection db structure
it also:
- handle the migration of old data structure into the new one while 
- setting a flag in db that the data has been migrated

the v2 implementation uses an array to hold all the protection data per validator that look like this
[ latest epoch written, source epoch, signing root, source epoch, signing root , source epoch, signing root...]

during the proposed tests by @nisdas I have found a race condition that can cause slashable offences to be signed by the validator

**Which issues(s) does this PR fix?**

Part of #7196

data interchange format
https://hackmd.io/@sproul/Bk0Y0qdGD#Implementations

### CODE REVIEW

File: validator/client/attest_protect.go [hunk 1]
```diff
@@ -66,37 +75,51 @@ func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.Ind
...
 	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
-
 	// Previously pruned, we should return false.
-	if int(targetEpoch) <= int(history.LatestEpochWritten)-int(wsPeriod) {
+	lew, err := history.GetLatestEpochWritten(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get latest epoch written from encapsulated data")
+		return false
+	}
+	if int(targetEpoch) <= int(lew)-int(wsPeriod) {
```
Comment (rauljordan) [inline validator/client/attest_protect.go:89]: is it safe to downcast this? We should keep uint64 and instead do an overflow check

File: validator/client/attest_protect.go [hunk 2]
```diff
@@ -106,31 +129,60 @@ func isNewAttSlashable(history *slashpb.AttestationHistory, sourceEpoch, targetE
...
+func safeTargetToSource(ctx context.Context, history *kv.EncHistoryData, targetEpoch uint64) *kv.HistoryData {
 	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
-	if targetEpoch > history.LatestEpochWritten || int(targetEpoch) < int(history.LatestEpochWritten)-int(wsPeriod) {
-		return params.BeaconConfig().FarFutureEpoch
+	lew, err := history.GetLatestEpochWritten(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get latest epoch written from encapsulated data")
+		return nil
+	}
+	if targetEpoch > lew || int(targetEpoch) < int(lew)-int(wsPeriod) {
```
Comment (rauljordan) [inline validator/client/attest_protect.go:176]: same here

File: validator/db/iface/interface.go [hunk 3]
```diff
@@ -25,4 +26,8 @@ type ValidatorDB interface {
 	// Attester protection related methods.
 	AttestationHistoryForPubKeys(ctx context.Context, publicKeys [][48]byte) (map[[48]byte]*slashpb.AttestationHistory, error)
 	SaveAttestationHistoryForPubKeys(ctx context.Context, historyByPubKey map[[48]byte]*slashpb.AttestationHistory) error
+
+	//new attestation store methods
```
Comment (rauljordan) [inline validator/db/iface/interface.go:30]:
```suggestion
	// New attestation store methods
```

File: validator/db/iface/interface.go [hunk 4]
```diff
@@ -25,4 +26,8 @@ type ValidatorDB interface {
...
+	AttestationHistoryNewForPubKeys(ctx context.Context, publicKeys [][48]byte) (map[[48]byte]*kv.EncHistoryData, error)
```
Comment (rauljordan) [inline validator/db/iface/interface.go:31]: Can we call these v2 instead?

File: validator/db/kv/attestation_history_new.go [hunk 5]
```diff
@@ -7,6 +7,7 @@ import (
 	"github.com/pkg/errors"
```
Comment (rauljordan) [inline validator/db/kv/attestation_history_new.go:7]: Shall we rename these to v2 to be more semantic?

File: validator/client/attest_protect.go (hunk 1 above)
Comment (shayzluf) [reply validator/client/attest_protect.go:89]: thanks

File: validator/client/attest_protect.go (hunk 2 above)
Comment (shayzluf) [reply validator/client/attest_protect.go:176]: thanks

File: validator/db/iface/interface.go (hunk 4 above)
Comment (shayzluf) [reply validator/db/iface/interface.go:31]: replaced

File: validator/db/kv/attestation_history_new.go (hunk 5 above)
Comment (shayzluf) [reply validator/db/kv/attestation_history_new.go:7]: renamed, looks much better

File: validator/client/attest_protect.go [hunk 6]
```diff
@@ -106,31 +130,63 @@ func isNewAttSlashable(history *slashpb.AttestationHistory, sourceEpoch, targetE
...
 
 // safeTargetToSource makes sure the epoch accessed is within bounds, and if it's not it at
-// returns the "default" FAR_FUTURE_EPOCH value.
-func safeTargetToSource(history *slashpb.AttestationHistory, targetEpoch uint64) uint64 {
+// returns the "default" nil value.
+func safeTargetToSource(ctx context.Context, history *kv.EncHistoryData, targetEpoch uint64) *kv.HistoryData {
 	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
-	if targetEpoch > history.LatestEpochWritten || int(targetEpoch) < int(history.LatestEpochWritten)-int(wsPeriod) {
-		return params.BeaconConfig().FarFutureEpoch
+	lew, err := history.GetLatestEpochWritten(ctx)
```
Comment (nisdas) [inline validator/client/attest_protect.go:172]: can we have a better name then `lew` ? its a bit vague

File: validator/client/attest_protect.go [hunk 7]
```diff
@@ -66,37 +75,52 @@ func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.Ind
...
+	if !hd.IsEmpty() && !bytes.Equal(signingRoot[:], hd.SigningRoot) {
 		return true
 	}
 
 	// Check if the new attestation would be surrounding another attestation.
 	for i := sourceEpoch; i <= targetEpoch; i++ {
-		// Unattested for epochs are marked as FAR_FUTURE_EPOCH.
-		if safeTargetToSource(history, i) == farFuture {
+		// Unattested for epochs are marked as (*kv.HistoryData)(nil).
+		historyBoundry := safeTargetToSource(ctx, history, i)
```
Comment (nisdas) [inline validator/client/attest_protect.go:107]:
```suggestion
		historyBoundary := safeTargetToSource(ctx, history, i)
```

File: validator/client/attest_protect.go [hunk 8]
```diff
@@ -66,37 +75,52 @@ func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.Ind
...
+		if historyBoundry == (*kv.HistoryData)(nil) {
```
Comment (nisdas) [inline validator/client/attest_protect.go:108]:
```suggestion
		if historyBoundary == (*kv.HistoryData)(nil) {
```

File: validator/client/attest_protect.go [hunk 9]
```diff
@@ -66,37 +75,52 @@ func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.Ind
...
 			continue
 		}
-		if history.TargetToSource[i%wsPeriod] > sourceEpoch {
+
+		if historyBoundry.Source > sourceEpoch {
```
Comment (nisdas) [inline validator/client/attest_protect.go:112]:
```suggestion
		if historyBoundary.Source > sourceEpoch {
```

File: validator/client/attest_protect.go [hunk 10]
```diff
@@ -106,31 +130,63 @@ func isNewAttSlashable(history *slashpb.AttestationHistory, sourceEpoch, targetE
...
+	}
+	if lew >= wsPeriod && targetEpoch < lew-wsPeriod { //Underflow protected older then weak subjectivity check.
+		return nil
+	}
+	hd, err := history.GetTargetData(ctx, targetEpoch%wsPeriod)
+	if err != nil {
+		log.WithError(err).Errorf("Could not get target data for target epoch: %d", targetEpoch)
+		return nil
+	}
+	if hd.IsEmpty() {
```
Comment (nisdas) [inline validator/client/attest_protect.go:188]: this check should be performed by the caller of the function, return nil here makes no difference in the event `hd` is nil anyway.

File: validator/db/kv/attestation_history_v2.go [hunk 11]
```diff
@@ -42,81 +43,108 @@ func (hd EncHistoryData) assertSize() error {
...
+
+func emptyHistoryData() *HistoryData {
+	h := &HistoryData{Source: params.BeaconConfig().FarFutureEpoch, SigningRoot: bytesutil.PadTo([]byte{}, 32)}
+	return h
+}
+
+// NewAttestationHistoryArray creates a new encapsulated attestation history byte array
+// sized by the latest epoch written.
+func NewAttestationHistoryArray(target uint64) *EncHistoryData {
+	en := make(EncHistoryData, latestEpochWrittenSize+(target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize+historySize)
```
Comment (nisdas) [inline validator/db/kv/attestation_history_v2.go:64]: please refactor this line, 4 different operations are happening here at once.

File: validator/client/attest_protect.go (hunk 10 above)
Comment (shayzluf) [reply validator/client/attest_protect.go:188]: thanks

File: validator/db/kv/attestation_history_v2.go (hunk 11 above)
Comment (shayzluf) [reply validator/db/kv/attestation_history_v2.go:64]: refactored

File: validator/client/attest_protect.go [hunk 12]
```diff
@@ -21,13 +22,21 @@ func (v *validator) preAttSignValidations(ctx context.Context, indexedAtt *ethpb
...
+		return err
+	}
+	if ok && isNewAttSlashable(ctx, attesterHistory, indexedAtt.Data.Source.Epoch, indexedAtt.Data.Target.Epoch, sr) {
 		if v.emitAccountMetrics {
 			ValidatorAttestFailVec.WithLabelValues(fmtKey).Inc()
 		}
 		return errors.New(failedPreAttSignLocalErr)
 	} else if !ok {
-		log.WithField("publicKey", fmtKey).Debug("Could not get local slashing protection data for validator")
+		v.attesterHistoryByPubKeyLock.Lock()
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply validator/client/attest_protect.go]: moved to `AttestationHistoryForPubKeysV2`

**Comment by shayzluf** [comment]
@nisdas confirmed it works in runtime

File: validator/client/attest_protect.go [hunk 13]
```diff
@@ -21,13 +22,18 @@ func (v *validator) preAttSignValidations(ctx context.Context, indexedAtt *ethpb
 	v.attesterHistoryByPubKeyLock.RLock()
 	attesterHistory, ok := v.attesterHistoryByPubKey[pubKey]
```
Comment (nisdas) [inline validator/client/attest_protect.go:23]: This just holds a reference to the history, if another routine is also writing it at the same time, you will end up with conflicting histories.   The reference isn't threadsafe.

Comment (shayzluf) [reply validator/client/attest_protect.go:23]: implemented the change
will test as you suggested as soon as i get victor to stop running his keys

File: validator/client/attest_protect.go [hunk 14]
```diff
@@ -41,17 +47,23 @@ func (v *validator) preAttSignValidations(ctx context.Context, indexedAtt *ethpb
...
 }
 
-func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.IndexedAttestation, pubKey [48]byte) error {
+func (v *validator) postAttSignUpdate(ctx context.Context, indexedAtt *ethpb.IndexedAttestation, pubKey [48]byte, signingRoot [32]byte) error {
 	fmtKey := fmt.Sprintf("%#x", pubKey[:])
 	v.attesterHistoryByPubKeyLock.Lock()
+	defer v.attesterHistoryByPubKeyLock.Unlock()
 	attesterHistory, ok := v.attesterHistoryByPubKey[pubKey]
 	if ok {
-		attesterHistory = markAttestationForTargetEpoch(attesterHistory, indexedAtt.Data.Source.Epoch, indexedAtt.Data.Target.Epoch)
```
Comment (shayzluf) [inline validator/client/attest_protect.go:49]: race condition found here. if few threads were trying to vote with the same key they would all pass through the preAttSignValidations as they all read the same attesterHistory and when they get to postAttSignUpdate  there are no checks to block it

**Review by rauljordan — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

**Pull request merged**